snapcraft/plugins/qbs.py

```diff
@@ -0,0 +1,157 @@
+"""
+import os
+import shutil
+import snapcraft
+
+
+class QbsPlugin(snapcraft.BasePlugin):
+
+    @classmethod
+    def schema(cls):
+        schema = super().schema()
+        schema['properties']['qbs-options'] = {
+            'type': 'array',
+            'minitems': 1,
+            'uniqueItems': True,
+            'items': {
+                'type': 'string',
+            },
+            'default': [],
+        }
+
+        schema['properties']['qbs-profile'] = {
+            'enum': ['gcc', 'clang'],
+            'default': 'gcc',
+        }
+
+        schema['properties']['qbs-build-variant'] = {
+            'enum': ['debug', 'release'],
+            'default': 'release',
+        }
+
+        schema['properties']['qt-version'] = {
+            'enum': ['nil', 'qt4', 'qt5'],
+        }
+
+        schema['required'].append('qt-version')
+
+        return schema
+
+    @classmethod
+    def get_pull_properties(cls):
+        # Inform Snapcraft of the properties associated with pulling. If these
+        # change in the YAML Snapcraft will consider the pull step dirty.
+        return ['qt-version', 'qbs-profile']
+
+    @classmethod
+    def get_build_properties(cls):
+        # Inform Snapcraft of the properties associated with building. If these
+        # change in the YAML Snapcraft will consider the build step dirty.
+        return ['qbs-options', 'qbs-build-variant']
+
+    def __init__(self, name, options, project):
+        super().__init__(name, options, project)
+        self.build_packages.extend(['qbs', 'gcc', 'clang'])
+
+        if self.options.qt_version == 'qt5':
+            self.build_packages.extend(['qt5-qmake', 'qtbase5-dev'])
+        elif self.options.qt_version == 'qt4':
+            self.build_packages.extend(['qt4-qmake', 'libqt4-dev'])
+        elif self.options.qt_version != 'nil':
+            raise RuntimeError('Unsupported Qt version: {!r}'.format(
+                self.options.qt_version))
+
+    def build(self):
+        super().build()
+
+        env = self._build_environment()
+
+        # Setup the toolchains, there will only be gcc or clang by default
+        self.run(['qbs', 'setup-toolchains', '--detect'], env=env)
+
+        build_profile = ""
+
+        if self.options.qt_version == 'nil':
+            build_profile = self.options.qbs_profile
+        else:
+            # a unique'ish name for snap builds. Hopefully this shouldn't clash
+            # with any other local profiles. Each profile should look something
+            # like: profiles.snapcraft-qbs-qt5-clang
+            build_profile = 'snapcraft-qbs-{}-{}'.format(
+                self.options.qt_version,
+                self.options.qbs_profile)
+
+            # Find full qmake path as qbs setup-qt requires it
+            qmake = shutil.which('qmake')
+            if qmake is None:
+                # Let's just check it's not in a common location
+                qmake = os.path.join('usr', 'bin', 'qmake')
```

Comment:
I don't see a case where qmake is in /usr/bin, but which can't find it.
Do you think this conditional is necessary?